### PR TITLE
[SPARK-23503][SS] Enforce sequencing of committed epochs for Continuous Execution

### PR DESCRIPTION
**Pull request by spaced4ndy**
## What changes were proposed in this pull request?

Made changes to EpochCoordinator so that it enforces a commit order. In case a message for epoch n is lost and epoch (n + 1) is ready for commit before epoch n is, epoch (n + 1) will wait for epoch n to be committed first.

## How was this patch tested?

Existing tests in ContinuousSuite and EpochCoordinatorSuite.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/EpochCoordinator.scala [hunk 1]
```diff
@@ -145,18 +149,42 @@ private[continuous] class EpochCoordinator(
...
+        lastCommittedEpoch = epoch
+
+        // Commit subsequent epochs that are waiting to be committed.
+        var nextEpoch = lastCommittedEpoch + 1
+        while (epochsWaitingToBeCommitted.contains(nextEpoch)) {
+          val nextEpochCommits =
+            partitionCommits.collect { case ((e, _), msg) if e == nextEpoch => msg }
+          logDebug(s"Committing epoch $nextEpoch.")
+          writer.commit(nextEpoch, nextEpochCommits.toArray)
+          query.commit(nextEpoch)
```
Comment (jose-torres) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/EpochCoordinator.scala]: There's a bit of duplicated logic here - helper methods would probably be nice.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/EpochCoordinator.scala [hunk 2]
```diff
@@ -145,18 +149,42 @@ private[continuous] class EpochCoordinator(
...
+        for (k <- partitionCommits.keys.filter { case (e, _) => e < lastCommittedEpoch }) {
+          partitionCommits.remove(k)
+        }
+        for (k <- partitionOffsets.keys.filter { case (e, _) => e < lastCommittedEpoch }) {
+          partitionOffsets.remove(k)
+        }
+      } else {
+        logDebug(s"Epoch $epoch has received commits from all partitions" +
+          s"and is waiting for epoch ${epoch - 1} to be committed first.")
+        epochsWaitingToBeCommitted.add(epoch)
```
Comment (jose-torres) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/EpochCoordinator.scala]: Maybe swap the order of the if else. I'd forgotten what the condition was for after scrolling down here.

**Review by jose-torres — COMMENTED** [review]
The general approach looks correct to me.

**Comment by jose-torres** [comment]
LGTM, once #20983 is committed and we can pull in the tests from that PR.

**Comment by efimpoberezkin** [comment]
@tdas @jose-torres Hi, I tested this with https://github.com/apache/spark/pull/20983, can we run jenkins tests now?

**Comment by jose-torres** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/EpochCoordinator.scala [hunk 3]
```diff
@@ -137,30 +137,65 @@ private[continuous] class EpochCoordinator(
   private val partitionOffsets =
     mutable.Map[(Long, Int), PartitionOffset]()
 
+  private var lastCommittedEpoch = startEpoch - 1
+  // Remembers epochs that have to wait for previous epochs to be committed first.
+  private val epochsWaitingToBeCommitted = mutable.HashSet.empty[Long]
```
Comment (jose-torres) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/EpochCoordinator.scala]: This is orthogonal to the current PR, but I realized that both this and the commits/offsets maps are unbounded queues. We probably should introduce some SQLConf for the maximum epoch backlog, and report an error when too many stack up. I'll file a JIRA ticket for this.

**Comment by felixcheung** [comment]
Jenkins, ok to test

**Comment by efimpoberezkin** [comment]
@felixcheung hi, could you merge this?

**Review by tdas — APPROVED** [review]
This looks good, just requires a little bit more docs.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/EpochCoordinator.scala [hunk 4]
```diff
@@ -137,30 +137,65 @@ private[continuous] class EpochCoordinator(
...
+          partitionCommits.remove(k)
+        }
+        for (k <- partitionOffsets.keys.filter { case (e, _) => e < lastCommittedEpoch }) {
+          partitionOffsets.remove(k)
+        }
       }
     }
   }
 
+  private def findCommitsForEpoch(epoch: Long): Iterable[WriterCommitMessage] = {
```
[the remark this answers is not in the document]
Comment (efimpoberezkin) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/EpochCoordinator.scala]: @tdas done

**Comment by tdas** [comment]
jenkins retest this please.

**Comment by efimpoberezkin** [comment]
@tdas Hi, could you give me some advice on how to resolve this error? I'm not sure how this PR could cause this, especially considering it was passing tests before and no changes were made to it since then. Besides failed jenkins test says "It is not a test it is a sbt.testing.SuiteSelector" in description.

**Comment by jose-torres** [comment]
retest this please

**Comment by jose-torres** [comment]
(I agree that your PR isn't responsible here, there's a known problem with that suite.)

**Comment by efimpoberezkin** [comment]
@tdas Hi, seems like it's good to go. Could you merge this when you have time?

**Comment by tdas** [comment]
Jenkins retest this please.

**Comment by tdas** [comment]
@efimpoberezkin I will merge it. Let me do another round of tests.

**Comment by efimpoberezkin** [comment]
retest this please

**Comment by efimpoberezkin** [comment]
@tdas can't start tests

**Comment by tdas** [comment]
jenkins retest this please.

**Comment by jose-torres** [comment]
ugh that flaky kafka test. It's already reported, and I've been looking into it this week albeit with little luck.

**Comment by jose-torres** [comment]
retest this please

**Comment by tdas** [comment]
I merged this to master. Thanks. And sorry for the delay.

**Pull request closed without merging**